api.c

```diff
@@ -0,0 +1,1055 @@
+                break;
+            case 708:
+                opt = atoi(optarg);
+                if (opt > 0)
+                    bar_periodic_step = opt;
+                break;
+            case 709:
+                arg = optarg;
+                if (sscanf(arg, "%31[^:]:%31[^:]", bar_x_expr, bar_y_expr) < 1) {
+                    apierr("bar-position must be a single number or of the form x:y with a max length of 31\n");
+                }
+                break;
+            case 710:
+                bar_count = atoi(optarg);
+                if (bar_count > MAX_BAR_COUNT || bar_count < MIN_BAR_COUNT) {
+                    apierr("bar-count must be between %d and %d\n", MIN_BAR_COUNT, MAX_BAR_COUNT);
+                }
+                break;
+            case 711:
+                arg = optarg;
+                if (sscanf(arg, "%31s", bar_width_expr) != 1) {
+                    apierr("missing argument for bar-total-width\n");
+                }
+                break;
+
+            case 998:
+                image_raw_format = strdup(optarg);
+                break;
+            default:
+                break;
+        }
+    }
+
+    if (image_path != NULL) {
+        if (!is_directory(image_path)) {
+            pthread_mutex_lock(&redraw_mutex);
+            img = load_image(image_path, image_raw_format);
+            pthread_mutex_unlock(&redraw_mutex);
+        }
+        free(image_path);
+    }
+    free(image_raw_format);
+}
+
+static void init_str(str_t *str, size_t size) {
+    str->value = (char*) malloc(size);
+    memset(str->value, 0, size);
+    str->size = size;
+}
+static char** str_destruct(arr_t arr) {
+    size_t str_size = sizeof(str_t) * arr.size;
+    char **str = (char**) malloc(sizeof(str_t) * arr.size);
+    memset(str, 0, str_size);
+
+    for (int i = 0; i < arr.size; i++) {
+        str[i] = strdup(arr.items[i].value);
+    }
+    return str;
+}
+static void str_push_c(str_t *str, char c) {
+    int len_str = strlen(str->value);
+    if (len_str + 2 >= str->size) {
+        str->size = (len_str + 1) * 2;
+        str->value = (char *) realloc(str->value, str->size);
+    }
+    str->value[len_str] = c;
+    str->value[len_str+1] = '\0';
+}
+
+static void push_back(arr_t *arr, str_t str) {
+    arr->items = (str_t *) realloc(arr->items, sizeof(str_t) * (arr->size + 2));
+
+    str_t toadd;
+    init_str(&toadd, 0);
+    toadd.value = strdup(str.value);
+    toadd.size = str.size;
+
+    arr->items[arr->size] = toadd;
+    arr->size++;
+}
+
+static void init_str_arr(arr_t *arr) {
+    str_t null;
+    init_str(&null, 0);
+    arr->items = (str_t *) malloc(0);
+    arr->items[0] = null;
+    arr->size = 0;
+}
+
+/**
+ * Arguments parser based on "stringtoargcargv.cpp" by Bernhard Eder
+ * @see https://web.archive.org/web/20121030075237/http://bbgen.net/blog/2011/06/string-to-argc-argv
+ */
+static int parse_args(arg_data *data, char str[]) {
+    char c;
+    int i = 0;
+    str_t curr_arg;
```

Comment:
Can we refactor this function to not use str_t and arr_t? I think we can make this file simpler by removing srt_t and arr_t related functions and do mallocs/reallocs only (or even better, VLAs, if possible).

Reply:
Yeah, I just added them to make the `parse_args` much easier, but if any refactor is made, then nice!